### DISCUSSION TITLE
Goofy ahh PauseState.hx addon error

### DISCUSSION THREAD

**Discussion by THEkatinamicrowave** [CodenameCrew/CodenameEngine#302] [Q&A]
I don't know what to put as the parameters of the game over substate
```haxe
case "Lowtiergod Yourself 😂":
	openSubState(new GameOverSubstate());
	close();
```

**Comment by r6915ee** [comment]
The parameters should be according to this:

```haxe
openSubState(new GameOverSubstate(x:Float, y:Float, character:String = "bf-dead", player:Bool = true, gameOverSong:String = "gameOver", lossSFX:String = "gameOverSFX", retrySFX:String = "gameOverEnd"));
```

I haven't tested this, but likely `x` and `y` are the HaxeFlixel positions of your character. `character`, of course, is the Codename character to display. `player`, as far as I'm aware, might probably be if the regular BF is replaced. `gameOver` is what plays after the intro of the animation, `lossSFX` is the sound effect to play when getting a game over (e.g. BF dropping his mic), and `gameOverEnd` is the sound effect that plays when retrying from the Game Over screen.

**Reply by THEkatinamicrowave** [reply]
CONTENT WARNING: I might just literally be insane
--------------------

I was trying to find where the player character name is stored in the hx files and I FINALLY found it after digging through PlayState. I found this at line 49:
```haxe
public static var SONG:ChartData;
```
so then I tried to find what ChartData is and found it at line 12:
```haxe
import funkin.backend.chart.ChartData;
```
so then I went to ChartData and found this
```haxe
typedef ChartData = {
	public var strumLines:Array<ChartStrumLine>;
	public var events:Array<ChartEvent>;
	public var meta:ChartMetaData;
	public var codenameChart:Bool;
	public var stage:String;
	public var scrollSpeed:Float;
	public var noteTypes:Array<String>;

	public var ?fromMods:Bool;
}
```
character names are in strumLines so I went there and found this:
```haxe
typedef ChartStrumLine = {
	var characters:Array<String>;
	var type:ChartStrumLineType;
	var notes:Array<ChartNote>;
	var position:String;
	var ?visible:Null<Bool>;
	var ?strumPos:Array<Float>;
	var ?strumScale:Float;
	var ?scrollSpeed:Float;
	var ?vocalsSuffix:String;

	var ?strumLinePos:Float; // Backwards compatability
}
```
cool, I found character types; it's ChartStrumLineType. Now let's look there:
```haxe
enum abstract ChartStrumLineType(Int) from Int to Int {
	/**
	 * STRUMLINE IS MARKED AS OPPONENT - WILL BE PLAYED BY CPU, OR PLAYED BY PLAYER IF OPPONENT MODE IS ON
	 */
	var OPPONENT = 0;
	/**
	 * STRUMLINE IS MARKED AS PLAYER - WILL BE PLAYED AS PLAYER, OR PLAYED AS CPU IF OPPONENT MODE IS ON
	 */
	var PLAYER = 1;
	/**
	 * STRUMLINE IS MARKED AS ADDITIONAL - WILL BE PLAYED AS CPU EVEN IF OPPONENT MODE IS ENABLED
	 */
	var ADDITIONAL = 2;
}
```
And now I'm stuck. How do I display the name of a player character if it's stored as a number? Also yeah it doesn't accept
```haxe
PlayState.SONG.strumLines.type.PLAYER
```
so now I'm kinda out of options.

TL;DR: I don't know where to find the correct stuff

**Reply by r6915ee** [reply]
The `character` parameter when making a new `GameOverSubstate` itself is actually just a string that contains the character's JSON and XML file name. This is the character file that gets used for the game over animation. It would simply be the same as what has been inserted into that parameter.

**Reply by THEkatinamicrowave** [reply]
Yeah, I'm currently trying to find the x:Float and y:Float values.

**Reply by r6915ee** [reply]
I guess that is perhaps where the player character is positioned that? I'm not sure what case you are trying to use this for, so I'm assuming it's something similar to the regular game over screen.

**Reply by THEkatinamicrowave** [reply]
Basically, I'm opening GameOverSubstate.hx using the `openSubstate()` function, and that requires certain parameters to be filled in in order to open.
```haxe
openSubState(new GameOverSubstate();
```
I looked inside GameOverSubstate and found these parameters
```haxe
public function new(x:Float, y:Float, character:String = "bf-dead", player:Bool = true, gameOverSong:String = "gameOver", lossSFX:String = "gameOverSFX", retrySFX:String = "gameOverEnd")
{
	super();
	this.x = x;
	this.y = y;
	this.characterName = character;
	this.player = player;
	this.gameOverSong = gameOverSong;
	this.lossSFXName = lossSFX;
	this.retrySFX = retrySFX;
}
```
I decided to try looking for "this" to see if maybe that could lead me somewhere. After finding nothing, I went to the Haxeflixel Documentary and found that "this" is literally just the current instance so I got confused.

I looked a bit further down and found this little thing hiding beneath what I already found
```haxe
		var event = EventManager.get(GameOverCreationEvent).recycle(x, y, characterName, player, gameOverSong, gameOverSongBPM,lossSFXName, retrySFX);
		gameoverScript.call('create', [event]);

		x = event.x;
		y = event.y;
		characterName = event.character;
		player = event.player;
		gameOverSong = event.gameOverSong;
		gameOverSongBPM = event.bpm;
		lossSFXName = event.lossSFX;
		retrySFX = event.retrySFX;
		
```
So which parameters do I use? Because that might be the whole problem

**Reply by r6915ee** [reply]
It's the same parameters included at the root of the thread.

**Reply by THEkatinamicrowave** [reply]
ah ok
so then just comes the question of how do I fetch the player character from PlayState (I assume it's in PlayState) and then the rest of the stuff

**Reply by r6915ee** [reply]
Again, this has not been tested, but this is the closest thing I can come up with in limited time:

`PlayState.boyfriend.character`

**Reply by THEkatinamicrowave** [reply]
bruh I just checked PlayState and found this :|
![image](https://github.com/FNF-CNE-Devs/CodenameEngine/assets/89645987/3320e03c-65b4-4490-bbea-5ee52dffeaa4)
I'll see if a good ol' `PlayState.gameOver(boyfriend);` will do the trick so we don't have to find all the variable stuff

**Reply by THEkatinamicrowave** [reply]
yeah I just checked `public function gameOver()` this is gonna work for sure
![image](https://github.com/FNF-CNE-Devs/CodenameEngine/assets/89645987/7a832aae-ba74-4606-8201-6aa64394832f)
